### PR TITLE
Idle schedule evm task

### PR DESCRIPTION
**Pull request by zjb0807**
Closes: #612 #1459

### CODE REVIEW

File: modules/evm/src/lib.rs [hunk 1]
```diff
@@ -890,18 +909,26 @@ impl<T: Config> Pallet<T> {
...
+			)
 		})?;
 
 		// this should happen after `Accounts` is updated because this could trigger another updates on
 		// `Accounts`
 		frame_system::Pallet::<T>::dec_providers(&address_account)?;
 
-		Ok(size)
+		// estimate weight
+		let weight = (ContractStorageSizes::<T>::get(address)
```
Comment (xlc) [inline modules/evm/src/lib.rs:927]: I don't think it is useful to charge additional weight here. The tx fee will be charged based on max gas and refund unused but it cannot charge more if more weights are returned.

The additional removal work is done on idle block time, so it is ok for them to be free. We may also only refund x% of the deposit.

File: modules/evm/src/lib.rs [hunk 2]
```diff
@@ -1494,3 +1529,91 @@ impl<T: Config + Send + Sync> SignedExtension for SetEvmOrigin<T> {
 		Ok(())
 	}
 }
+
+#[derive(Clone, RuntimeDebug, PartialEq, Encode, Decode, TypeInfo)]
+pub enum EvmTask<T: Config> {
+	// TODO: update
```
Comment (xlc) [inline modules/evm/src/lib.rs:1535]: write an issue for those TODO

File: modules/evm/src/lib.rs [hunk 3]
```diff
@@ -1494,3 +1529,91 @@ impl<T: Config + Send + Sync> SignedExtension for SetEvmOrigin<T> {
...
+							finished: false,
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+impl<T: Config> From<EvmTask<T>> for () {
```
Comment (xlc) [inline modules/evm/src/lib.rs:1615]: this is for test only right? should add cfg flag

File: modules/support/src/lib.rs [hunk 4]
```diff
@@ -600,3 +601,25 @@ pub trait CallBuilder {
...
+pub trait DispatchableTask {
+	fn dispatch(self, weight: Weight) -> TaskResult;
+}
+
+/// Idle scheduler trait
+pub trait IdleScheduler<Task> {
+	fn schedule(task: Task) -> DispatchResult;
+}
+
+impl DispatchableTask for () {
```
Comment (xlc) [inline modules/support/src/lib.rs:615]: should use cfg to mark those for test only

File: primitives/src/task.rs [hunk 5]
```diff
@@ -16,18 +16,29 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 use frame_support::weights::Weight;
+use sp_runtime::DispatchResult;
+
+#[allow(dead_code)]
```
Comment (xlc) [inline primitives/src/task.rs:21]: is this still a dead code?

**Review by xlc — COMMENTED** [review]
now delete contract is not atomic action. is it possible to create a contract with same address of a deleting one? I think that's possible and we need to make sure that action will fail, instead of overriding storages.

**Comment by ermalkaleci** [comment]
we need to make sure it's not possible to deploy at the same address until everything is wiped out

File: modules/evm/src/lib.rs (hunk 2 above)
Comment (zjb0807) [reply modules/evm/src/lib.rs:1535]: #1591

**Comment by zjb0807** [comment]
> we need to make sure it's not possible to deploy at the same address until everything is wiped out

https://github.com/AcalaNetwork/Acala/blob/ba2e0cee5e6d78aa06ac40ef5dfa444f169efd4d/modules/evm/src/runner/state.rs#L580-L583

I checked the `storage_size` and remove it at
https://github.com/AcalaNetwork/Acala/blob/ba2e0cee5e6d78aa06ac40ef5dfa444f169efd4d/modules/evm/src/lib.rs#L1584 

I'll add a test case.

**Comment by zjb0807** [comment]
In order to avoid adding extra `storage_size`, delay deleting account to check the same address deployment problem

File: modules/evm/src/lib.rs [hunk 6]
```diff
@@ -1494,3 +1521,94 @@ impl<T: Config + Send + Sync> SignedExtension for SetEvmOrigin<T> {
...
+							"EvmTask::Remove: [from: {:?}, contract: {:?}, maintainer: {:?}, count: {:?}, result: {:?}]",
+							caller, contract, maintainer, count, res
+						);
+
+						// Remove account after all of the storages are cleared.
+						Accounts::<T>::take(contract);
+
+						TaskResult {
+							result: res,
+							used_weight: <T as frame_system::Config>::DbWeight::get()
```
Comment (xlc) [inline modules/evm/src/lib.rs:1582]: we should do some real benchmarking here to include the weights of refund and remove accounts

this can happen later

**Review by xlc — APPROVED** [review]

**Pull request merged**